tests/drivers/bc12/src/main.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @defgroup driver_bc12_tests
+ * @ingroup all_tests
+ * @{
+ * @}
+ */
+
+#include <zephyr/drivers/usb/usb_bc12.h>
+#include <zephyr/drivers/usb/emul_bc12_pi3usb9201.h>
+#include <zephyr/ztest.h>
+
+#include <zephyr/logging/log.h>
+
+#define LOG_MODULE_NAME test_bc12
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_INF);
+
+struct bc12_pi3usb9201_fixture {
+	const struct device *bc12_dev;
+	const struct emul *bc12_emul;
+	int callback_count;
+	bool disconnect_detected;
+	struct bc12_partner_state charging_partner;
+};
+
+static void bc12_test_result_cb(const struct device *dev, struct bc12_partner_state *state,
+				void *user_data)
+{
+	struct bc12_pi3usb9201_fixture *fixture = user_data;
+
+	fixture->callback_count++;
+
+	if (state) {
+		LOG_INF("callback: type %d, voltage %d, current %d", state->type, state->voltage,
+			state->current);
+		fixture->charging_partner = *state;
+	} else {
+		LOG_INF("callback: partner disconnect");
+		fixture->disconnect_detected = true;
+		fixture->charging_partner.type = BC12_TYPE_NONE;
+		fixture->charging_partner.current = 0;
+		fixture->charging_partner.voltage = 0;
+	}
+}
+
+ZTEST_USER_F(bc12_pi3usb9201, test_bc12_no_charging_partner)
+{
+	if (fixture->bc12_dev == NULL) {
+		ztest_test_skip();
+	}
+
+	zassert_true(device_is_ready(fixture->bc12_dev));
+
+	bc12_result_cb(fixture->bc12_dev, &bc12_test_result_cb, fixture);
```

Comment:
We call `bc12_result_cv` on each test. I think we ought to just add it to the setup after asserting on bc12_dev not being null and device_is_ready.